### ISSUE TITLE
[RFC] Dropping legacy networking mode

### ISSUE THREAD

**Issue by akerouanton** [moby/moby#47464]
### Description

During the last couple weeks, we had a number of discussions on what should be done to remove legacy networking mode, and more particularly about legacy links. As there were many proposals made by various maintainers, this RFC try to summarize the various ideas floating around.

- [Goals](#goals)
- [Consensus](#consensus)
- [Points of contention](#points-of-contention)
    1. [Reimplement legacy links in terms of DNS records](#reimplement-legacy-links-in-terms-of-DNS-records)
    2. [Disable ICC altogether](#disable-ICC-altogether)
    3. [Enable ICC](#enable-ICC)
- [Deprecation plan](#deprecation-plan)

## Goals

1. Deprecate and remove a feature marked as legacy for almost 10 years (finally!) ;
2. Reduce the complexity exposed to end-users and make it easier for them to transition from the default network to a custom network. In that regard, the default network should be a subset / "diminished version" of, or exactly the same as what custom networks provide. Currently, that's not the case because of the links feature having different semantics ;
3. Lastly, get rid of the current snowflake implementation used by the default network. This would reduce the complexity of libnetwork.

## Consensus

I think there're two things we all agree about:

1. `/etc/hosts`-based discovery should be let go and replaced by DNS-based discovery. It's recognized that this file isn't an API -- we don't mind breaking users who parse it ;
2. The current env vars mechanism is obsolete and should be removed. They're barely useful anyway since you've to parse `_NAME` env vars and know the port you want to reach to look up the right env vars:

```
$ docker run --rm -d --name=c0 -p 80:80 alpine top
$ docker run --rm -it --name=c1 --link=c0 alpine /bin/sh -c export
export C0_NAME='/c1/c0'
export C0_PORT='tcp://172.17.0.4:80'
export C0_PORT_80_TCP='tcp://172.17.0.4:80'
export C0_PORT_80_TCP_ADDR='172.17.0.4'
export C0_PORT_80_TCP_PORT='80'
export C0_PORT_80_TCP_PROTO='tcp'
```

Mostly as a side note (it probably requires a dedicated proposal), a replacement mechanism was briefly discussed. That mechanism could be used to solve other issues: #3778 (5th most upvoted networking issue, 25th overall) and #8427. This could take the form of:

1. A filesystem mount, allowing discoverability by walking the file tree. Another pro: it solves the notification issue (think Traefik use-case), and can be expanded to non-networking needs -- killing two birds with one stone. But there was some pushback on this design ;
2. Using SRV records. Cons:
  - The notification issue isn't solved ;
  - Thus (most probably) limited to ports discovery (eg. host port when publishing to an ephemeral port) ;
  - Ports don't have labels right now -- how to leverage RFC2782?

## Points of contention

The current point of contention is about what the default bridge should look like, and what should be the semantics of links (on that network).

### 1. Reimplement legacy links in terms of DNS records

The default network would be icc=false, and links would be used to declare which containers could reach each other.

From a user standpoint, the links semantics would still be unique to the default bridge, hardly hitting goals 2. and 3. However, we could change the semantics of links on custom networks to make it look like the current legacy semantics. But, this doesn't seem user-driven, making it hard to justify time investment and making the overall design more complex.

The new semantics would be:

- When icc=true, links are container-specific CNAME records ;
- When icc=false, same as icc=true plus they also define which containers can reach each other ;

From a technical standpoint, this is currently not easily feasible because non-legacy links are  implemented as container-specific CNAME records pointing to network-wide A records. When the DNS resolver is disabled, no A records are currently populated (**needs confirmation**).

### 2. Disable ICC altogether

Basically make the default bridge a north-south network -- ie. no inter-container connectivity, only a gateway to host's LAN/WAN. Legacy links should be replaced with custom networks, in line with what the docs currently advertise: https://docs.docker.com/network/links/.

There was some pushback on this idea as this would be a _massive_ breaking change -- this was already tried in the past and reverted.

### 3. Enable ICC

All containers connected to the default bridge can resolve and communicate with each other. No more legacy links -- links on the default bridge behave like on custom networks.

From a user standpoint, the default bridge network is just like any other bridge network. From a technical standpoint, no more snowflake implementation. This is probably the easiest solution, and the one which best fit goals 1-3.

This has the advantage of unifying Linux & Windows behaviors (the embedded DNS is enabled on Windows).

If users are unhappy about icc=true, we could still allow them to recreate the default bridge with whatever options they need, eg. like `docker_gwbridge`. There have been some discussions around unifying the way predefined networks are created / configured (eg. deprecate top-level daemon flags, unify the experience, etc...), so I'll keep those details for another proposal.

There was some pushback too but that's still too fuzzy for me, so I'll let anyone on the fence chime in. One potential blocker is that containers disconnected from Swarm networks end up being reattached to the default network -- but that looks like a silly bug warranting a proper fix (**need to double check this claim**).

## Deprecation plan

This mostly depends on what alternative we adopt, but it'd be a variation of something like:

1. Show a WARNING on `docker create` / `docker run` whenever links are used on the default bridge network for a release or two ;
2. Disable legacy networking / links by default, and add a feature flag to re-enable them if needed ;
4. After a few releases, get totally rid of them.

**Comment by sparrc** [comment]
Hello, I work at AWS ECS on the [ECS agent](https://github.com/aws/amazon-ecs-agent).

If I understand the RFC correctly, we would probably prefer option (3). From our perspective default ICC is probably OK (and actually that's the main reason [customers have asked us for user-defined networks](https://github.com/aws/containers-roadmap/issues/184) going on 8 years). If we determine that default ICC is a problem for our customers we could relatively easily create our own `ecs` "user-defined" network with ICC disabled and launch containers into that.

One point of clarification I was wondering: with option 3, does the `--link` flag still exist in perpetuity? It will just behave differently? (ie, no env var sharing)

For context, at ECS we do still offer ["links" as a feature](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definition_parameters.html#container_definition_network), and we still also launch containers into the default docker0 bridge network. If links completely went away (from my understanding this is option 2) we would have a big problem on our hands in terms of trying to support backward compatibility of ECS task definitions with the latest versions of docker engine.

Anyways just my 2 cents, thanks for publishing the RFC and communicating what the maintainers are thinking on this :)

**Comment by akerouanton** [comment]
Thanks @sparrc for the feedback.

> One point of clarification I was wondering: with option 3, does the --link flag still exist in perpetuity? It will just behave differently? (ie, no env var sharing)

The idea behind option 3 is to completely ditch the links feature. But it seems we now have a good reason not to do so.

I start to think the best option is to go with option 1. but make the links feature available to custom networks.

When users want to connect a frontend to multiple backends (say one reverse proxy like traefik, and multiple HTTP services), they're generally creating a single network. This topology has one major downside: all backends can now talk to each other. With the links feature made available to custom networks, this could be solved. In that case, icc would be false and links would be used to define ACLs and DNS records.

**Comment by thaJeztah** [comment]
> I start to think the best option is to go with option 1. but make the links feature available to custom networks.

Note that `--link` on custom networks was never deprecated, only the `--link` implementation on the default bridge (so the legacy implementation with env-var sharing etc).

**Comment by neersighted** [comment]
Right, but it never 'worked' the same way on custom networks (just serving as a form of alias); option 1 requires per-container-network-pair DNS resolution mapping so that the conditional visibility can be replicated.

**Comment by kszarlej** [comment]
Please make sure that you finally add the option to start the tasks in user-defined network on the host so:

- The bidirectional linking limitation goes away
- We can use DNS discovery with the container hostname
